src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAO.scala

```diff
@@ -302,13 +303,145 @@ class PostgresDirectoryDAO(protected val dbRef: DbReference,
+                from ${PetServiceAccountTable as pet}
+                where ${pet.email} = ${email}
+              union
+              select ${None}, ${None}, ${None}, ${None}, ${pol.name}, ${res.name}, ${srt.name}
+                from ${PolicyTable as pol}
+                join ${GroupTable as g}
+                on ${pol.groupId} = ${g.id}
+                join ${ResourceTable as res}
+                on ${res.id} = ${pol.resourceId}
+                join ${ResourceTypeTable as srt}
+                on ${res.resourceTypeId} = ${srt.id}
+                where ${g.email} = ${email}"""
+
+      val result = query.map(rs =>
+        SubjectConglomerate(
+          rs.stringOpt(1),
+          rs.stringOpt(2),
+          rs.stringOpt(3),
+          rs.stringOpt(4),
+          rs.stringOpt(5),
+          rs.stringOpt(6),
+          rs.stringOpt(7)
+        )).single().apply()
+
+      result.map(unmarshalSubjectConglomerate)
+    }.recoverWith {
+      case e: TooManyRowsException =>
+        IO.raiseError[Option[WorkbenchSubject]](new WorkbenchException(s"Database error: email $email refers to too many subjects."))
+    }
+  }
+
+  def loadPolicyEmail(policyId: FullyQualifiedPolicyId): IO[Option[WorkbenchEmail]] = {
+    runInTransaction { implicit session =>
+      val g = GroupTable.syntax
+      val pol = PolicyTable.syntax
+      val srt = ResourceTypeTable.syntax
+      val res = ResourceTable.syntax
+
+      import SamTypeBinders._
+
+      val query = samsql"""
+                     select ${g.email}
+                     from ${PolicyTable as pol}
+                     join ${GroupTable as g}
+                      on ${pol.groupId} = ${g.id}
+                     join ${ResourceTable as res}
+                      on ${res.id} = ${pol.resourceId}
+                     join ${ResourceTypeTable as srt}
+                      on ${res.resourceTypeId} = ${srt.id}
+                     where ${srt.name} = ${policyId.resource.resourceTypeName} and
+                      ${res.name} = ${policyId.resource.resourceId} and
+                      ${pol.name} = ${policyId.accessPolicyName}"""
+
+      query.map(rs => rs.get[WorkbenchEmail](g.email)).single().apply()
+    }
+  }
+
+  override def loadSubjectEmail(subject: WorkbenchSubject): IO[Option[WorkbenchEmail]] = {
+    subject match {
+      case subject: WorkbenchGroupName => loadGroupEmail(subject)
+      case subject: PetServiceAccountId => for {
+        petSA <- loadPetServiceAccount(subject)
+      } yield petSA.map(_.serviceAccount.email)
+      case subject: WorkbenchUserId => for {
+        user <- loadUser(subject)
+      } yield user.map(_.email)
+      case subject: FullyQualifiedPolicyId => loadPolicyEmail(subject)
+      case _ => throw new WorkbenchException(s"unexpected subject [$subject]")
+    }
+  }
+
+  override def loadSubjectEmails(subjects: Set[WorkbenchSubject]): IO[Stream[WorkbenchEmail]] = {
```

Comment:
Should this handle pets and policies?

Reply:
I had wondered this too but this is actually an exact copy-paste from `LdapDirectoryDAO`

Reply:
🤷‍♂